src/renderer/dx/DxFontRenderData.cpp

```diff
@@ -0,0 +1,758 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "precomp.h"
+
+#include "DxFontRenderData.h"
+
+static constexpr float POINTS_PER_INCH = 72.0f;
+static constexpr std::wstring_view FALLBACK_FONT_FACES[] = { L"Consolas", L"Lucida Console", L"Courier New" };
+static constexpr std::wstring_view FALLBACK_LOCALE = L"en-us";
+
+using namespace Microsoft::Console::Render;
+
+DxFontRenderData::DxFontRenderData(::Microsoft::WRL::ComPtr<IDWriteFactory1> dwriteFactory) noexcept :
+    _dwriteFactory(dwriteFactory),
+    _glyphCell{},
+    _lineMetrics({}),
+    _boxDrawingEffect{},
+    _boxDrawingEffectItalic{}
+{
+}
+
+[[nodiscard]] Microsoft::WRL::ComPtr<IDWriteTextAnalyzer1> DxFontRenderData::Analyzer() noexcept
+{
+    return _dwriteTextAnalyzer;
+}
+
+[[nodiscard]] Microsoft::WRL::ComPtr<IDWriteFontFallback> DxFontRenderData::GetSystemFontFallback()
+{
+    if (!_systemFontFallback)
+    {
+        ::Microsoft::WRL::ComPtr<IDWriteFactory2> factory2;
+        THROW_IF_FAILED(_dwriteFactory.As(&factory2));
+        factory2->GetSystemFontFallback(&_systemFontFallback);
+    }
+
+    return _systemFontFallback;
+}
+
+[[nodiscard]] til::size DxFontRenderData::GlyphCell() noexcept
+{
+    return _glyphCell;
+}
+
+[[nodiscard]] DxFontRenderData::LineMetrics DxFontRenderData::GetLineMetrics() noexcept
+{
+    return _lineMetrics;
+}
+
+[[nodiscard]] Microsoft::WRL::ComPtr<IDWriteTextFormat> DxFontRenderData::DefaultTextFormat() noexcept
+{
+    return _dwriteTextFormat;
+}
+
+[[nodiscard]] Microsoft::WRL::ComPtr<IDWriteFontFace1> DxFontRenderData::DefaultFontFace() noexcept
+{
+    return _dwriteFontFace;
+}
+
+[[nodiscard]] Microsoft::WRL::ComPtr<IBoxDrawingEffect> DxFontRenderData::DefaultBoxDrawingEffect() noexcept
+{
+    return _boxDrawingEffect;
+}
+
+[[nodiscard]] Microsoft::WRL::ComPtr<IDWriteTextFormat> DxFontRenderData::ItalicTextFormat() noexcept

```

Comment:
In case anyone's wondering, my approach of adding bold & bold-italic is just adding `BoldTextFormat`, `BoldFontFace`, `BoldItalicTextFormat` and `BoldItalicFontFace` here. And use them in `CustomTextLayout` accordingly.

Reply:
Better than what we have now I suppose. Kinda gross to have 4, but it is what it is.

Reply:
I don't like to have like 6 of these things, either. I see this PR as the beginning of a series refactoring work. I think we can find a better way to deal with them later based on this PR.